store/cache.go

```diff
@@ -0,0 +1,169 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package store
+
+import (
+	"crypto"
+	"encoding/hex"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+// downloadCache is the interface that a store download cache must provide
+type downloadCache interface {
+	// Get gets the given sha3 content and puts it into targetPath
+	Get(sha3_384, targetPath string) error
+	// Put adds a new file to the cache
+	Put(sourcePath string) error
+	// Lookup checks if the given sha3 content is available in the cache
+	Lookup(sha3_384 string) bool
+}
+
+// nullCache is cache that does not cache
+type nullCache struct{}
+
+func (cm *nullCache) Get(sha3_384, targetPath string) error {
+	return fmt.Errorf("cannot get items from the nullCache")
+}
+func (cm *nullCache) Put(sourcePath string) error { return nil }
+func (cm *nullCache) Lookup(sha3_384 string) bool { return false }
+
+// changesByReverseMtime sorts by the mtime of files
+type changesByReverseMtime []os.FileInfo
+
+func (s changesByReverseMtime) Len() int           { return len(s) }
+func (s changesByReverseMtime) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+func (s changesByReverseMtime) Less(i, j int) bool { return s[i].ModTime().After(s[j].ModTime()) }
+
+// cacheManager implements a downloadCache via content based hard linking
+type CacheManager struct {
+	cacheDir string
+	maxItems int
+}
+
+// NewCacheManager returns a new CacheManager with the given cacheDir
+// and the given maximum amount of items. The idea behind it is the
+// following algorithm:
+//
+// 1. When starting a download, check if it exists in $cacheDir
+// 2. If found, update its mtime, hardlink into target location, and
+//    return success
+// 3. If not found, download the snap
+// 4. On success, hardlink into $cacheDir/<digest>
+// 5. If cache dir has more than maxItems entries, remove oldest mtimes
+//    until it has maxItems
+//
+// The caching part is done here, the downloading happens in the store.go
+// code.
+func NewCacheManager(cacheDir string, maxItems int) *CacheManager {
+	return &CacheManager{
+		cacheDir: cacheDir,
+		maxItems: maxItems,
+	}
+}
+
+// Get gets the given sha3 content and puts it into targetPath
+func (cm *CacheManager) Get(sha3_384, targetPath string) error {
+	if !cm.Lookup(sha3_384) {
+		return fmt.Errorf("cannot find %q in %q", sha3_384, cm.cacheDir)
+	}
+	if err := os.Link(cm.path(sha3_384), targetPath); err != nil {
+		return err
+	}
+	return os.Chtimes(targetPath, time.Now(), time.Now())
+}
+
+// Put adds a new file to the cache
+func (cm *CacheManager) Put(sourcePath string) error {
+	// happens on e.g. `snap download` which runs as the user
+	if !osutil.IsWritable(cm.cacheDir) {
+		return nil
+	}
+
+	if err := os.MkdirAll(cm.cacheDir, 0700); err != nil && !os.IsExist(err) {
+		return err
+	}
+
+	sha3_384, err := cm.digest(sourcePath)
```

Comment:
As an optimization, should we accept this as an optional parameter, in case the call site already has the digest at hand? Haven't seen the rest of the code, but it would be awkward for the download logic to request adding something into the cache without first making sure it's correct.

Reply:
Yes, we can do that. You suggest that if cache.Put(file, hash) is called the Put() code trusts the caller and will not verify the hash again, correct? I feel slightly uneasy about this, my gut-feeling is that we should validate the hash but it might be slow.

Reply:
I was thinking about this one and you are right, the cache does not need to deal with hashes at all. I pushed 2982ccfe2a9292a280a229c5f943c2b66e7785f5 that essentially makes it use `cacheKeys` instead of hashes. The knowledge about hashes is already part of the store code and we don't really need it in cache (at least currently). I also simplified the code and removed "Lookup()" as its racy.

Reply:
Please let me know if you think this approach is ok, I can revert to the "smarter" cache if you think this simplfies it too much.